File: biblioteca-backend/src/main/java/com/azurian/library/service/AuthorService.java
package com.azurian.library.service;

import com.azurian.library.dto.request.AuthorRequest;
import com.azurian.library.dto.response.AuthorResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface AuthorService {
    Page<AuthorResponse> findAll(String search, Pageable pageable);
    AuthorResponse findById(Long id);
    AuthorResponse create(AuthorRequest request);
    AuthorResponse update(Long id, AuthorRequest request);
    void delete(Long id);
}

File: biblioteca-frontend/src/app/core/services/author.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Author, AuthorRequest } from '../models/author.model';
import { Page, PageParams } from '../models/page.model';

@Injectable({ providedIn: 'root' })
export class AuthorService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/authors`;

  getAll(params: PageParams = {}): Observable<Page<Author>> {
    let httpParams = new HttpParams()
      .set('page', params.page ?? 0)
      .set('size', params.size ?? 10)
      .set('sortBy', params.sortBy ?? 'lastName')
      .set('sortDir', params.sortDir ?? 'asc');

    if (params.search) {
      httpParams = httpParams.set('search', params.search);
    }

    return this.http.get<Page<Author>>(this.baseUrl, { params: httpParams });
  }

  getById(id: number): Observable<Author> {
    return this.http.get<Author>(`${this.baseUrl}/${id}`);
  }

  create(request: AuthorRequest): Observable<Author> {
    return this.http.post<Author>(this.baseUrl, request);
  }

  update(id: number, request: AuthorRequest): Observable<Author> {
    return this.http.put<Author>(`${this.baseUrl}/${id}`, request);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}
